### PR TITLE
`azurerm_synapse_sql_pool` - leverage newer version mssql

### PR DESCRIPTION
**Pull request by njuCZ**
follow up of PR #8095

### CODE REVIEW

File: azurerm/internal/services/mssql/parse/mssql.go [hunk 1]
```diff
@@ -29,6 +29,18 @@ type MsSqlDatabaseExtendedAuditingPolicyId struct {
 	ResourceGroup string
 }
 
+func NewMsSqlDatabaseID(resourceGroup, msSqlServer, name string) MsSqlDatabaseId {
```
Comment (WodansSon) [inline azurerm/internal/services/mssql/parse/mssql.go:32]: Should this be a more descriptive name?

```suggestion
func SynapseSqlPoolDatabaseID(resourceGroup, msSqlServer, name string) MsSqlDatabaseId {
```

File: azurerm/internal/services/mssql/parse/mssql.go [hunk 2]
```diff
@@ -29,6 +29,18 @@ type MsSqlDatabaseExtendedAuditingPolicyId struct {
...
+	return MsSqlDatabaseId{
+		ResourceGroup: resourceGroup,
+		MsSqlServer:   msSqlServer,
+		Name:          name,
+	}
```
Comment (WodansSon) [inline azurerm/internal/services/mssql/parse/mssql.go:37]: Shouldn't all of these values be checked and validated and return an error if they are not valid?

File: azurerm/internal/services/mssql/parse/mssql.go [hunk 3]
```diff
@@ -29,6 +29,18 @@ type MsSqlDatabaseExtendedAuditingPolicyId struct {
...
+}
+
+func (id MsSqlDatabaseId) ID(subscriptionId string) string {
```
Comment (WodansSon) [inline azurerm/internal/services/mssql/parse/mssql.go:40]: Same here, shouldn't the subscriptionId input be validated and return an error if it is null, empty, or malformed?

File: azurerm/internal/services/mssql/parse/mssql.go (hunk 2 above)
Comment (WodansSon) [inline azurerm/internal/services/mssql/parse/mssql.go:37]: Same here, shouldn't we expose a new `struct` with a more descriptive name?
```suggestion
	return SynapseSqlPollDatabaseId{
		ResourceGroup: resourceGroup,
		SynapseSqlPoolServer:   synapseSqlPoolServer,
		Name:          name,
	}
```

File: azurerm/internal/services/synapse/synapse_sql_pool_resource.go [hunk 4]
```diff
@@ -113,7 +112,7 @@ func resourceArmSynapseSqlPool() *schema.Resource {
 				Optional:     true,
 				Computed:     true,
 				ForceNew:     true,
-				ValidateFunc: postgresValidate.PostgresDatabaseCollation,
+				ValidateFunc: mssqlValidate.MsSqlDBCollation(),
```
Comment (WodansSon) [inline azurerm/internal/services/synapse/synapse_sql_pool_resource.go:115]: Is this collation specific to only synapse sql pool's or is it generic for all msSql databases? If it is for only synapse I would update the name to be reflective of that, something like `synapseSqlPoolDBCollation`

File: azurerm/internal/services/synapse/synapse_sql_pool_resource.go [hunk 5]
```diff
@@ -123,7 +122,7 @@ func resourceArmSynapseSqlPool() *schema.Resource {
 				ConflictsWith: []string{"restore"},
 				ValidateFunc: validation.Any(
 					validate.SqlPoolID,
-					sqlValidate.SqlDatabaseID,
+					mssqlValidate.MsSqlDatabaseID,
```
Comment (WodansSon) [inline azurerm/internal/services/synapse/synapse_sql_pool_resource.go:125]: Same here, is this for all Ms Sql Databases or is it just for synapse sql pool's? If it's just for pool's I would change the name here as well.

**Review by WodansSon — CHANGES_REQUESTED** [review]

**Comment by njuCZ** [comment]
hi @WodansSon, this is a rule that Tom commented: https://github.com/terraform-providers/terraform-provider-azurerm/pull/8095#discussion_r473871779 . This PR is to solve @jackofallops comments in last PR. mssql is a newer version than sql, so this PR change to reference mssql. The source_database_id format is either synapse_sql_pool id or sql database id. But the backend service only supports sql database id format, so need to convert the format

**Comment by WodansSon** [comment]
image

**Review by WodansSon — APPROVED** [review]

**Pull request merged**